enhancements/update/cvo-log-level-api.md

```diff
@@ -0,0 +1,327 @@
+---
+title: cvo-log-level-api
+authors:
+  - "@DavidHurta"
+reviewers:
+  - "@wking"
+  - "@LalatenduMohanty"
+  - "@petr-muller"
+approvers:
+  - "@wking"
+api-approvers:
+  - "@deads2k"
+  - "@bparees"
+  - "@JoelSpeed"
+creation-date: 2023-10-09
+last-updated: 2024-10-01
+tracking-link:
+  - https://issues.redhat.com/browse/OTA-923
+see-also:
+  - "None"
+replaces:
+  - "None"
+superseded-by:
+  - "None"
+---
+
+# CVO Log Level API
+
+## Summary
+
+This enhancement proposes to create a new OpenShift API to provide a simple
+method of dynamically changing the verbosity level of Cluster Version Operator's
+logs. There will be four log levels available. The lowest level being the
+current default log level being used by the Cluster Version Operator.
+
+## Motivation
+
+There is currently no way to easily change the verbosity of the [Cluster Version Operator (CVO)](https://github.com/openshift/cluster-version-operator)
+logs in a live cluster.
+
+It would be useful to provide functionality for the cluster administrators and
+OpenShift engineers to easily modify the log level to a desired level using an API
+[similarly as can be done for some other operators](https://github.com/openshift/api/blob/5852b58f4b1071fe85d9c49dff2667a9b2a74841/operator/v1/types.go#L67-L74).
+
+### User Stories
+
+* As an OpenShift administrator, I want to increase the log level of the CVO 
+  from the default level to more easily troubleshoot any potential issues regarding
+  the cluster or the CVO.
+* As an OpenShift administrator, I want to decrease the log level of the CVO
+  from a previously set higher log level to save storage space.
+* As an OpenShift engineer, I want to increase the log level of the CVO for
+  a CI run so that I can more easily troubleshoot any potential issues.
+
+### Goals
+
+* Add a user-facing API for controlling the verbosity of the CVO logs.
+
+### Non-Goals
+
+* Change the default logging verbosity of the Cluster Version Operator.
+* Allow users to set a lower than the current default logging level.
+
+## Proposal
+
+This enhancement proposes to create a new `CustomResourceDefinition` (CRD) 
+called `clusterversionoperators.operator.openshift.io`. The CRD will be part 
+of the [`github.com/openshift/api/operator/v1`][github.com/openshift/api/operator/v1]
+package. A `ClusterVersionOperator` resource will be used to configure the 
+CVO. The configuration, as of now, will only contain the knob to modify the CVO 
+log level. A `ClusterVersionOperator` resource named `cluster` will be added to
+the OCP payload. This resource will act as a singleton configuration resource
```

Comment:
I wanted to be sure of the current state, so I have looked closer.

A payload contains a bunch of `operator.openshift.io` resources, for example:
```
$ grep -sRil "^apiVersion: operator.openshift.io/v1$" /tmp/release/release-manifests/
/tmp/release/release-manifests/0000_50_cloud-credential-operator_01-operator-config.yaml
/tmp/release/release-manifests/0000_50_cluster-authentication-operator_02_config.cr.yaml
/tmp/release/release-manifests/0000_10_config-operator_01_operator.cr.yaml
/tmp/release/release-manifests/0000_50_cluster-csi-snapshot-controller-operator_02_operator_cr.yaml
/tmp/release/release-manifests/0000_20_etcd-operator_01_operator.cr.yaml
/tmp/release/release-manifests/0000_20_kube-apiserver-operator_01_operator.cr.yaml
/tmp/release/release-manifests/0000_25_kube-controller-manager-operator_01_operator.cr.yaml
/tmp/release/release-manifests/0000_25_kube-scheduler-operator_02_operator.cr.yaml
/tmp/release/release-manifests/0000_50_cluster-kube-storage-version-migrator-operator_06_operatorconfig.yaml
/tmp/release/release-manifests/0000_50_cluster-openshift-apiserver-operator_01_operator.cr.yaml
/tmp/release/release-manifests/0000_50_cluster-openshift-controller-manager-operator_03_config.cr.yaml
/tmp/release/release-manifests/0000_50_cluster-storage-operator_06_operator_cr-hypershift.yaml
/tmp/release/release-manifests/0000_50_cluster-storage-operator_06_operator_cr.yaml
/tmp/release/release-manifests/0000_50_console-operator_01-operator-config.yaml
/tmp/release/release-manifests/0000_50_insights-operator_04-operator-cr.yaml
/tmp/release/release-manifests/0000_50_service-ca-operator_03_operator.cr.yaml
```

I have looked at how exactly it functions in a cluster. These resources (at least the ones I have inspected) have an annotation `release.openshift.io/create-only: "true"`. [Resources with such an annotation should be created if they do not exist, but should not be updated if they already exist](https://github.com/openshift/cluster-version-operator/blob/master/lib/resourceapply/interface.go#L9-L13). This annotation would be used for the proposed `ClusterVersionOperator` `cluster` resource (in case the CR is part of the OCP payload) so that the CVO only creates the resource.

> I believe typically they are either generated by the operator when not present, or by the installer at cluster install time.

It's possible that there are configurations created by their operators as well. I am not aware of them, but it could be an alternative solution to this enhancement. Let me reach out to the HyperShift team regarding the HyperShift aspect.

Reply:
Sounds good, lets resolve this thread then